README.md
<p align="center">

  <img src="https://img.shields.io/badge/TypeScript-007ACC?style=flat&logo=typescript&logoColor=white" />
  <img src="https://img.shields.io/badge/JavaScript-F7DF1E?style=flat&logo=javascript&logoColor=black" />
  <img src="https://img.shields.io/badge/Node.js-43853D?style=flat&logo=node.js&logoColor=white" />

</p>

# @bitarray/es6

![GitHub package.json dynamic](https://img.shields.io/github/package-json/keywords/swiing/bitarray)

A BitArray class for easy and native-like operations on sequences of bits.

## Rationale

The library implements [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer)s behind the hood. This allows for:

- **optimal memory usage**: each bit is effectively coded in just one bit in memory, as opposed to being coded as, e.g. booleans or integers in regular arrays.

- **efficient bitwise operations**: operations treat 32 bits at once, as opposed to dealing with entries one by one in regular arrays.

> :bulb: It comes at the cost that accessing a single bit at a given index will be slightly more expensive than accessing the value stored in a regular array: indeed, the interesting bit needs to be extracted once a Uint32 value has been found first.
>
> Having said that, in the very vast majority of cases, any difference either way should be unnoticeable; and you may very well opt to use this library for the sole reason that it provides **a convenient [native-like interface](#usage)** for your needs.

## Compatibility

![compatibility](https://img.shields.io/badge/compatibility-%3E%3D%20ES6-orange?style=flat)

The library relies on the [Proxy](https://caniuse.com/?search=Proxy) object, which is an ES6 (aka ES2015) feature. It can NOT be polyfilled (to the extent it is used by the library).

_Note: standard `TypedArray` is also a feature of ecmascript ES6._

## Installation

[![npm version](https://badge.fury.io/js/@bitarray%2Fes6.svg)](https://badge.fury.io/js/@bitarray%2Fes6)

```sh
npm install @bitarray/es6
```

or

```sh
yarn add @bitarray/es6
```

## Usage

Usage is same as for any standard typed array. You may check the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray) for details. This is because the `BitArray` object extends [@bitarray/typedarray](https://github.com/swiing/Bit-TypedArray). Check it for details.

It adds [bitwise operations](#Bitwise-operations) on top.

### Instantiating

```js
import BitArray from "@bitarray/es6"

const bits = BitArray.from("11001010");

const otherbits = new BitArray(12); // 12 bits, the values of which default to zero.
```

### Bitwise operations

```js
bits.count; // === 4; number of bits set to 1.

// same as bits.and(otherbits)
(bits)['&'](otherbits); // an instance holding a sequence of eight zeros

// same as bits.or(otherbits)
(bits)['|'](otherbits); // 11001010

// same as bits.xor(otherbits)
(bits)['^'](otherbits); // 11001010
```

### Iterating

```js
for (let i=0; i<bits.length; i++)
  // do something with bits[i]

bits.forEach((val, i, arr) => { /* do something */ });

for (let i in bits)
  // do something with bits[i]

for (let bit of bits)
  // do something with bit
```

### Indexes & values

```js
Object.keys(bits);    // [0, 1, 2, 3, 4, 5, 6, 7]
Object.values(bits);  // [1, 1, 0, 0, 1, 0, 1, 0]
Object.entries(bits); // [["0", 1], ["1", 1], ["2", 0], ["3", 0], ...]
```

## License

[![license](https://img.shields.io/github/license/swiing/BitArray)](https://github.com/swiing/BitArray/blob/main/LICENSE)


CHANGELOG.md
# @bitarray/es6 ChangeLog

## v1.2.0

_2023-09-02_

- fix: make it of "module" type ([9712c08](https://github.com/swiing/BitArray/commit/9712c0894fb00cc33ec27bceadd6d431504ae832))
- fix: set correct path for typings field in package.json ([#30](https://github.com/swiing/BitArray/pull/30)) (thanks [@yuhr](https://github.com/yuhr)!)
- chore: drop nodejs v14, add nodejs v20 ([afbff84](https://github.com/swiing/BitArray/commit/afbff84d5535f7605fe890232ed22d8eb243afba))

## v1.1.2

_2022-08-30_

- fix: correct exports path in package.json

## v1.1.1

_2022-08-07_

- fix: file extension in import
- fix: files list in tsconfig
- fix: typescript type error in example
- build: bump typescript version

## v1.1.0

_2022-08-07_

- chore: add .vscode/
- style: add and run prettier
- chore: remove dist/ from repo
- refactor: encode/decode
- feat: add encoding/decoding of bit array to string ([#1](https://github.com/swiing/BitArray/pull/1)) (thanks [@rollie42](https://github.com/rollie42)!)
- build: move creation of package.json to postbuild
- chore: add dependabot.yml
- build: add ci workflow
- build: bump dependencies versions

## v1.0.2

_2022-04-20_

- build: make typescript happy
- chore: update license to MIT

## v1.0.1

_2022-04-03_

- docs: improve readme

## v1.0.0

_2022-04-03_

- build: prepare for publication (esm+cjs)

## v0.2.0

_2022-04-02_


test/test.ts
// coloring in node
const NodeFailure = '\x1b[31m', // red
  NodeSuccess = '\x1b[32m', // green
  NodeInfo = '\x1b[37m'; // white

// coloring in browser
const BrowserFailure = 'color:red;',
  BrowserSuccess = 'color:green;',
  BrowserInfo = 'color:darkgrey;font-size: x-large;';

const log =
  typeof window === 'object'
    ? // in browser
      function (success: boolean, text: string, tab: string) {
        success == true
          ? console.log('%c' + tab + text, BrowserSuccess)
          : success == false
          ? console.log('%c' + tab + text, BrowserFailure)
          : console.log('%c' + tab + text, BrowserInfo);
      }
    : // in node
      function (success: boolean, text: string, tab: string) {
        success == true
          ? console.log(NodeSuccess + tab + text + NodeInfo)
          : success == false
          ? console.log(NodeFailure + tab + text + NodeInfo)
          : console.log(tab + text);
      };

export default function test(suite: object, tab = '') {
  let success: boolean;

  // for( let [desc, success] of Object.entries( suite ) ) // es2017
  for (let desc of Object.keys(suite)) {
    success = suite[desc];

    if (typeof success === 'boolean') log(success, desc, tab);
    else {
      log(undefined, desc, tab);
      test(success, tab + '\t');
    }
  }
}


test/index.ts
import suite from './suite';
import test from './test';

test(suite);


test/suite.js
import BitArray from "../../es6/dist/bitarray.js";

const len = 42; // choose whatever value

// we fill randomly, but we could choose fixed values, or whatever,
const arr1 = new Array( len      ).fill(false).map( x => Math.random() > 0.5 )
// arr2 is arbitrarily longer here. Could also be shorter, or same size
const arr2 = new Array( len + 10 ).fill(false).map( x => Math.random() > 0.5 )

const sample1 = BitArray.from( arr1 );
const sample2 = BitArray.from( arr2 );

// matches the format of BitArray.toSting()
function toString( arr ) {
  return arr.map( Number )
              // add space between each byte
              .map( (b,i) => (i+1)%8 ? b : b+" " )
              .join("")
              // remove possible end space
              .trim();
}


/** suite 1 */
const instantiating = {
  "is instanceof BitArray": sample1 instanceof BitArray
};

/** suite 2 */
const properties = {
  ".count": sample1.count == arr1.reduce( (acc,bit) => acc+Number(bit), 0 )
};


/** suite 3 */
const binary_operations = (()=>{

  const shortestArr = arr2.length>arr1.length ? arr1 : arr2;

  return {
    "&":   (sample1)['&'](sample2) .toString() === toString( shortestArr.map((_,i)=> arr1[i] && arr2[i] )),
    
    "|":   (sample1)['|'](sample2) .toString() === toString( shortestArr.map((_,i)=> arr1[i] || arr2[i] ))
        // make sure we don't have bits set beyond the length of the bitarray
        && (sample1)['|'](sample2) .count == arr1.reduce( (acc,bit,i) => acc+(Number(arr1[i])|Number(arr2[i])), 0 ),
        
    "^":   (sample1)['^'](sample2) .toString() === toString( shortestArr.map((_,i)=> arr1[i] ^  arr2[i] ))
        && (sample1)['^'](sample2) .count == arr1.reduce( (acc,bit,i) => acc+(Number(arr1[i])^Number(arr2[i])), 0 ),
    
    "or" : (sample1).or  (sample2) .toString() === toString( shortestArr.map((_,i)=> arr1[i] || arr2[i] )),
    "and": (sample1).and (sample2) .toString() === toString( shortestArr.map((_,i)=> arr1[i] && arr2[i] )),
    "xor": (sample1).xor (sample2) .toString() === toString( shortestArr.map((_,i)=> arr1[i] ^  arr2[i] ))
  }

})();

export default {
  instantiating,
  properties,
  binary_operations
};



src/bitarray.ts
/**
 *  An ES6 BitArray class for easy operations on sequences of bits.
 *
 *  @author swiing
 */

/**
 *  It aims at ease of use.
 *
 *  With regards to performance, it remains to be experimented/proven
 *  that it actually performs better than a regular array of booleans/numbers
 *  (at least in the specific context where it is used).
 *  Presumably, accessing values at indexes (array[i]) is slow, because
 *  it goes via a proxy. On the other hand, logical operations should be
 *  really fast. So it may depend on how much of those operations are used.
 *  Unless performance is a critical part, it should be ok for most cases
 *  anyway (it may be possible to increase performance by minimizing
 *  accesses to the proxy handlers - FFS if needed).
 *
 *  Lastly, it is very memory-efficient, as each value is coded in 1 bit,
 *  as opposed to booleans being less efficiently coded.
 *  (see https://dev.to/shevchenkonik/memory-size-of-javascript-boolean-3mlj)
 *
 */

import BitTypedArray from '@bitarray/typedarray';

import { base64MIMEChars, base64UrlChars } from './alphabet.js';

const alphabetLengthErrorMsg = "Alphabet's length must be a power of 2";

// I could leverage _views from @bitarray/typedarray, or create a new one here.
// import { _views } from "./bit-typedarray.js";
const _views = new WeakMap<ArrayBuffer, Uint32Array>();

function getViewer(instance: BitArray) {
  let viewer = _views.get(instance.buffer);
  if (!viewer) {
    _views.set(instance.buffer, (viewer = new Uint32Array(instance.buffer)));
  }
  return viewer;
}

// This is an extension of BitTypedArray with features (methods,getters)
// not available in native TypedArray's, essentially, bit operations.
export default class BitArray extends BitTypedArray {
  and: (bitArray: BitArray) => BitArray;
  or: (bitArray: BitArray) => BitArray;
  xor: (bitArray: BitArray) => BitArray;

  static from: (source: Iterable<any>) => BitArray;
  static of: (...args) => BitArray;

  /**
   * returns the number of 'true' entries
   */
  // this is meant to be performant
  // inspired by https://github.com/bramstein/bit-array/blob/master/lib/bit-array.js
  get count(): number {
    let count = 0;
    let view = getViewer(this);
    for (let i = 0; i < this.length / 32; i++) {
      let x = view[i];
      // See: http://bits.stephan-brumme.com/countBits.html for an explanation
      x = x - ((x >> 1) & 0x55555555);
      x = (x & 0x33333333) + ((x >> 2) & 0x33333333);
      x = x + (x >> 4);
      x &= 0xf0f0f0f;
      count += (x * 0x01010101) >> 24;
    }
    return count;
  }

  // Implementation note 1: I once used the maximum length of the two operands (for | and for ^),
  // defaulting to 'false' for the operand with "missing" values. However, it is arguable that
  // this is a good default assumption (it depends on context) so I finally prefer to go
  // conservative, and compute only when there is actual data available (i.e. not assuming
  // 'false' otherwise).
  // As a positive side effect, it also simplifies implementation (but this is not the driver).

  // Implementation note 2: I could have a very simple implementation doing e.g.
  // ret[i] = this[i] && bitArray[i], for every index
  // However, this would go through the array bit by bit. Instead, I go 4 bytes by 4 bytes (Uint32).
  // This is where improved performance is expected.

  /**
   *
   * @param bitArray
   * @returns a BitArray instance with the logical 'or' applied to this and the passed parameter.
   * The length of the resulting BitArray is the minimum of the length of the two operands.
   */
  ['|'](bitArray: BitArray): BitArray {
    const ret = new BitArray(Math./*max*/ min(this.length, bitArray.length));
    const retView = new Uint32Array(ret.buffer);
    _views.set(ret.buffer, retView);
    const thisView = getViewer(this);
    const bitArrayView = getViewer(bitArray);

    // let offset = 0;
    // -1 needed for the case len == 32*n
    let offset = ((ret.length - 1) >> 5) + 1; // divide 32, and add 1 because of next i--.

    // for( /**/;
    //      // -1 needed for the case Math.min() == 32*n.
    //      offset < ( ret.length /*== Math.min( this.length, bitArray.length )*/ - 1 >> 5 ) + 1;
    //      offset++ )
    while (offset--)
      // note: ret is a proxy;
      retView[offset] = thisView[offset] | bitArrayView[offset];

    // Needed only if length of the returned value would be the max length of the two operands
    // for( /**/;
    //      // offset < ( Math.max( this.length, bitArray.length ) >> 5 ) + 1;
    //      // (tbc) same as:
    //      offset < ret.length >> 5;
    //      offset++ )
    //    _views.get( ret.buffer )[ offset ] = _views.get( ( this.length <= bitArray.length ? bitArray : _targets.get( this ) ).buffer )[ offset ];

    // when the two operands have a different length, and since bitwise
    // operations treat by bunch of 32 bits, we may have set bits to 1
    // beyond the length of ret. We apply a mask to set them back to zero
    // (so that other operations like .count are correct)
    retView[ret.length >> 5] &=
      (1 << (ret.length & 31)) - 1; /* modulo 32 , -1*/

    return ret;
  }

  /**
   *
   * @param bitArray
   * @returns a BitArray instance with the logical 'and' applied to this and the passed parameter.
   * The length of the resulting BitArray is the minimum of the length of the two operands.
   */
  ['&'](bitArray: BitArray): BitArray {
    const ret = new BitArray(Math.min(this.length, bitArray.length));
    const retView = new Uint32Array(ret.buffer);
    _views.set(ret.buffer, retView);
    const thisView = getViewer(this);
    const bitArrayView = getViewer(bitArray);

    // -1 needed for the case len == 32*n
    let offset = ((ret.length - 1) >> 5) + 1; // divide 32, and add 1 because of next i--.

    while (offset--) retView[offset] = thisView[offset] & bitArrayView[offset];

    return ret;
  }

  /**
   *
   * @param bitArray
   * @returns a BitArray instance with the logical 'xor' applied to this and the passed parameter.
   * The length of the resulting BitArray is the minimum of the length of the two operands.
   */
  ['^'](bitArray: BitArray): BitArray {
    const ret = new BitArray(Math.min(this.length, bitArray.length));
    const retView = new Uint32Array(ret.buffer);
    _views.set(ret.buffer, retView);
    const thisView = getViewer(this);
    const bitArrayView = getViewer(bitArray);

    // -1 needed for the case len == 32*n
    let offset = ((ret.length - 1) >> 5) + 1; // divide 32, and add 1 because of next i--.

    while (offset--) retView[offset] = thisView[offset] ^ bitArrayView[offset];

    // when the two operands have a different length, and since bitwise
    // operations treat by bunch of 32 bits, we may have set bits to 1
    // beyond the length of ret. We apply a mask to set them back to zero
    // (so that other operations like .count are correct)
    retView[ret.length >> 5] &=
      (1 << (ret.length & 31)) - 1; /* modulo 32 , -1*/

    return ret;
  }

  /**
   *
   * @param alphabet a set of n characters to use to encode the BitArray; alphabet.length must be a power of 2 (2, 4, 8, etc)
   * The more characters in the set, the more compact the resulting output will be
   * @returns a string encoded using the provided character set (e.g., base64 encoding can be achieved with this)
   */
  encode(alphabet: string): string {
    const log2 = Math.log2(alphabet.length);

    if (log2 < 1 || log2 % 1 !== 0) {
      throw new RangeError(alphabetLengthErrorMsg);
    }

    const ret = [];

    let val = 0;
    let valLen = 0;
    for (const b of this) {
      valLen++;
      val <<= 1;
      val += b;

      if (valLen === log2) {
        ret.push(alphabet[val]);
        valLen = val = 0;
      }
    }

    if (valLen !== 0) {
      val <<= log2 - valLen;
      ret.push(alphabet[val]);
    }

    return ret.join('');
  }

  /**
   *
   * @param alphabet a set of n characters to use to encode the BitArray; alphabet.length must be a power of 2 (2, 4, 8, etc),
   * and should generally match the set used in the original encoding
   * @param encodedString an encoded string built with encode
   * @returns a BitArray of the encodedString decoded using alphabet
   */
  static decode(encodedString: string, alphabet: string): BitArray {
    const log2 = Math.log2(alphabet.length);

    if (log2 < 1 || log2 % 1 !== 0) {
      throw new RangeError(alphabetLengthErrorMsg);
    }

    const pad = (s: string) => '0'.repeat(log2 - s.length) + s;

    const charMap = {}; // maps each character to its integral value
    for (var i = 0; i < alphabet.length; i++) {
      charMap[alphabet[i]] = pad(i.toString(2));
    }

    const deserialized = Array.from(encodedString)
      .map((c) => {
        if (!(c in charMap)) {
          throw new RangeError('Invalid character found in encoded string');
        }
        return charMap[c];
      })
      .join('');

    return BitArray.from(deserialized);
  }

  // Convenience specializations for encoding base64MIME and base64Url
  encodeBase64MIME() {
    return this.encode(base64MIMEChars);
  }
  static decodeBase64MIME(encodedString: string) {
    return BitArray.decode(encodedString, base64MIMEChars);
  }

  encodeBase64Url() {
    return this.encode(base64UrlChars);
  }
  static decodeBase64Url(encodedString: string) {
    return BitArray.decode(encodedString, base64UrlChars);
  }
}

// create aliases
BitArray.prototype.and = BitArray.prototype['&'];
BitArray.prototype.or = BitArray.prototype['|'];
BitArray.prototype.xor = BitArray.prototype['^'];


example/index.js
import BoolArray from '../dist/bitarray.js';
import { log as _, logHeader as _$ } from "./util.js";


const len = 33; // choose any integer value

_$("Randomly initializing an array");

  const randomArray = new BoolArray(len);
  for( let i=0, bool; i<len; i++)
      randomArray[i] = (Math.random() > 0.5);

  _( "instanceof == BoolArray", randomArray instanceof BoolArray )

  _( "array == ", randomArray);


_$("properties");

  _( ".count", randomArray.count);


_$("Bitwise operations");

  const sample1 = BoolArray.from([true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,true ]
                        .concat([ true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,true ]));

  const sample2 = BoolArray.from([false,true,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,false,
                                  false,true,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,true ]
                        .concat([ true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,false,
                                  true,false,false,true,false,false,false,true ]));

  _( "sample1 = ", sample1.toString() );
  _( "sample2 = ", sample2.toString() );
  _( "s1 | s2 = ", (sample1)['|'](sample2) .toString() );
  _( "s1 ^ s2 = ", (sample1)['^'](sample2) .toString() );
  _( "s1 & s2 = ", (sample1)['&'](sample2) .toString() );


example/index.ts
import BoolArray from '../src/bitarray.js';
import { log as _, logHeader as _$ } from './util.js';

import { bit } from '@bitarray/typedarray';

const len = 33; // choose any integer value

_$('Randomly initializing an array');

const randomArray = new BoolArray(len);
for (let i = 0; i < len; i++)
  randomArray[i] = (Math.random() > 0.5) as unknown as bit;

_('instanceof == BoolArray', randomArray instanceof BoolArray);

_('array == ', randomArray);

_$('properties');

_('.count', randomArray.count);

_$('Bitwise operations');

const sample1 = BoolArray.from(
  [
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    true,
  ].concat([
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    true,
  ])
);

const sample2 = BoolArray.from(
  [
    false,
    true,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    false,
    true,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    true,
  ].concat([
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    false,
    true,
    false,
    false,
    true,
    false,
    false,
    false,
    true,
  ])
);

_('sample1 = ', sample1.toString());
_('sample2 = ', sample2.toString());
_('s1 | s2 = ', sample1['|'](sample2).toString());
_('s1 ^ s2 = ', sample1['^'](sample2).toString());
_('s1 & s2 = ', sample1['&'](sample2).toString());
